### PR TITLE
New Builder: Joyent Triton

### PR DESCRIPTION
**Pull request by jen20**
This is a builder for Joyent's Triton system. It was originally at jen20/packer-builder-triton, and subsequently at joyent/packer-builder-triton on GitHub.

### CODE REVIEW

**Comment by justindthomas** [comment]
Made a couple of tweaks in https://github.com/mitchellh/packer/pull/3815 to get this to build on my Mac.

File: builder/triton/ssh.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+		return "", err
+	}
+
+	return machine, nil
+}
+
+func sshConfig(state multistep.StateBag) (*ssh.ClientConfig, error) {
+	config := state.Get("config").(Config)
+
+	signer, err := ssh.ParsePrivateKey([]byte(config.Comm.SSHPrivateKey))
```
Comment (siepkes) [inline builder/triton/ssh.go]: I encountered the following error:

```
[DEBUG] Error getting SSH config: Error parsing SSH private key: ssh: no key found
```

`config.Comm.SSHPrivateKey` holds the path to the file containing the private key. Not the raw private key it self. The function `ParseRawPrivateKey` which is later called expects the raw private key in a byte[].

The following change fixes the problem:
```
privateKeyBytes, err := ioutil.ReadFile(config.Comm.SSHPrivateKey)
if err != nil {
	return nil, fmt.Errorf("Unable to read private key from file '%s': %s", config.Comm.SSHPrivateKey, err)
}

signer, err := ssh.ParsePrivateKey(privateKeyBytes)
```

**Comment by siepkes** [comment]
I encountered a problem with `ssh.go` treating the file location of the SSH private key as the private key. I've started a conversation in the "Files changed" section at the `ssh.go` file.

**Comment by siepkes** [comment]
On a somewhat unrelated note; What would it take (besides the obvious rebasing) to get this functionality merged?

**Comment by jen20** [comment]
Re the key: does it have a paraphrase?

**Comment by jen20** [comment]
Ah ignore, sorry I only read the last comment. I think we need to fix this up to support key path or literals and also use of SSH agents. I'll try to do so this week.

**Comment by siepkes** [comment]
I also found another (unrelated, non showstopper) issue. If the "create-from-vm" job fails on Triton Packer will still display success:

```
2016/11/06 10:06:29 machine readable: triton,artifact []string{"0", "files-count", "0"}
2016/11/06 10:06:29 machine readable: triton,artifact []string{"0", "end"}
2016/11/06 10:06:29 ui: --> triton: Image was created: 679b297f-8f03-4561-9ea4-c5f194bcce3c
2016/11/06 10:06:29 waiting for all plugin processes to complete...
--> triton: Image was created: 679b297f-8f03-4561-9ea4-c5f194bcce3c
```

The job is marked as failed in Triton. It failed because my Triton imgapi wasn't configured correctly. Thats obviously not something packer has to deal with but I think packer should do something with the failure it receives from Triton.

**Comment by siepkes** [comment]
@jen20 How can I help to get this merged? 

Currently we maintain an internal fork of Packer with this functionality. We rebase against master to stay updated however for obvious reasons I would rather invest time in getting this merged then investing time in maintaining a fork.

**Comment by rickard-von-essen** [comment]
If someone fixes the above issues and rebases I can at least start to review the code. (Witch is one step to get this in)

**Comment by siepkes** [comment]
@rickard-von-essen Great! So which issues need to be addressed from your point of view? If I skim this pull request correctly I recognize the following open issues which need to be addressed:

* "Propper" SSH support.
* Packer dealing correctly with Triton errors.
* Rebase

Let me know if this is a complete list or if I'm missing something.

@jen20 I don't want to "hijack" your hard work, are you okay with me stepping in on this? I don't think I can update this pull request, right? Should I create a new one with the above issues addressed? Is that okay with you?

**Comment by rickard-von-essen** [comment]
That sounds about right. Ensure there are some tests, preferably both unit
and acceptance, at least unit. And docs.

On Dec 11, 2016 10:30 AM, "Jasper Siepkes" <notifications@github.com> wrote:

@rickard-von-essen <https://github.com/rickard-von-essen> Great! So which
issues need to be addressed from your point of view? If I skim this pull
request correctly I recognize the following open issues which need to be
addressed:

   - "Propper" SSH support.
   - Packer dealing correctly with Triton errors.
   - Rebase

Let me know if this is a complete list or if I'm missing something.

@jen20 <https://github.com/jen20> I don't want to "hijack" your hard work,
are you okay with me stepping in on this? I don't think I can update this
pull request, right? Should I create a new one with the above issues
addressed? Is that okay with you?

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/mitchellh/packer/pull/3441#issuecomment-266272025>, or mute
the thread

**Comment by rickard-von-essen** [comment]
Superseded by #4325

**Pull request closed without merging**